### PR TITLE
Restructure Tools Menu

### PR DESCRIPTION
**Pull request by uliska**
The Tools menu is becoming too long IMO, so I think we should introduce additional submenus, even if that requires one more level to access some options. This PR introduces the following changes:

* New "Code Formatting" submenu with reformat, indent etc.  
  I have some ideas to add more "linting" commands that would further extend that list of commands, which is an additional incentive to move them in a submenu
* Submenus for Tools Panels

To be discussed:

* Naming of the panels submenus
* organization of panels in these submenus
* should the remaining tools also be moved to submenus, if so, what should they be called like
* If that is accepted the mnemonics for the panels can be reconsidered because they have to be unique only in their submenus
* Should the section from "Pitch" to "Quick Remove" also be moved to a submenu? That would be consistent but may move some regularly used commands one level deeper
* Should "Update with convert-ly" be moved to either the new "Code Formatting" submenu or to the LilyPond menu?

### CODE REVIEW

**Comment by uliska** [comment]
image

**Comment by uliska** [comment]
I have now also moved the musical transformations to a new submenu. I had hesitated to do so because it will add one layer to access the functions, *but* I had the idea to make that whole new submenu also available in the editor's and all viewers' context menus, so when you right-click on an item in the score the context menu will have an item `Musical Transformations=>` available (note: this is not included in this PR).

image

**Comment by fedelibre** [comment]
I thought just few days ago that the Tools menu was too long. I like the new restructure.

Just a few comments for now:

- what about adding "Layout Control Options" to the Viewers submenu? It's basically an advanced music  viewer.
-  then I would group the two MIDI tools in a MIDI submenu

**Comment by fedelibre** [comment]
> Should "Update with convert-ly" be moved to either the new "Code Formatting" submenu or to the LilyPond menu?

I think it should be moved to the LilyPond menu, as it's about LilyPond syntax version.

**Comment by fedelibre** [comment]
This comment is not much related to this PR.
IMO the Documents menu is not that useful and it's stealing space which might be assigned to something else.

I usually use Alt+arrows to move from one document to another.
If I have lots of files open, then I'd rather open the Documents pane, which opens on the left by default. This is what you expect from most editors; you don't have this list in a dedicated menu.

**Comment by uliska** [comment]
Am 22. Juli 2018 12:59:57 MESZ schrieb Federico Bruni <notifications@github.com>:
>This comment is not much related to this PR.
>IMO the Documents menu is not that useful and it's stealing space which
>might be assigned to something else.
>
>I usually use Alt+arrows to move from one document to another.
>If I have lots of files open, then I'd rather open the Documents pane,
>which opens on the left by default. This is what you expect from most
>editors; you don't have this list in a dedicated menu.

I know that some people including Wilbert feel pretty strong about that menu,and while the documents *panel* may be the more traditional place I think the problem is that it *permanently* uses space. Alt+D is a very fast way of accessing the list of documents.

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by fedelibre** [comment]
> I've done so, but I'm not 100% convinced. convert-ly operates on the currently opened document and not on the LilyPond installation (like e.g. the available fonts). This would make it more likely to be placed next to the new "Code formatting" and "Musical transformations" submenus in Tools?

Well, also the Engrave commands in that menu operate on the currently opened document.

Updating the syntax of a file is not exactly a "code formatting" operation.
But I think I see a point in keeping it in the Tools menu:

- LilyPond menu is mainly about producing output. You may even call it Engrave...
- Tools menu is mainly about changing the input file, so it's where you'd expect to find a tool like convert-ly

I think I changed my mind and I'd rather place it in Code formatting submenu.

**Comment by uliska** [comment]
I'd rather place it below the two submenus but within the same group of separators. The commands in Code Formatting have in common that they don't change the content in any way.

**Comment by uliska** [comment]
... which would be the following (uncommitted yet):

image

File: frescobaldi_app/panelmanager.py [hunk 1]
```diff
@@ -45,33 +47,49 @@ def __init__(self, mainwindow):
...
-            self.loadPanel("objecteditor.ObjectEditor")
+            self.loadPanel("objecteditor.ObjectEditor", "coding")
         self.createActions()
 
         # make some default arrangements
         mainwindow.tabifyDockWidget(self.musicview, self.docbrowser)
         mainwindow.tabifyDockWidget(self.musicview, self.svgview)
 
-    def loadPanel(self, name):
+        print(self._submenus)
```
Comment (dliessi) [inline frescobaldi_app/panelmanager.py]: @uliska Please, remember to delete debugging lines like this one before committing.

**Comment by dliessi** [comment]
I like this change very much and I also like the idea of having the "musical transformation" submenu in the context menu.
Regarding the position of "update with convert-ly", I agree with
> I'd rather place it below the two submenus but within the same group of separators.

**Comment by dliessi** [comment]
Shouldn't also "show available fonts" be moved to the "tools" menu?

File: frescobaldi_app/panelmanager.py (hunk 1 above)
Comment (uliska) [reply frescobaldi_app/panelmanager.py]: Arggh

**Comment by uliska** [comment]
> Shouldn't also "show available fonts" be moved to the "tools" menu?

I will do so once #1075 is merged.

**Comment by uliska** [comment]
This is the current state:

image

**Comment by uliska** [comment]
Yet another modification: I've merged a few commands in a new `Tools=>Directories` submenu.

For the reasoning I copy the commit message:

> This warrants another submenu because (when we may finally merge
the Git code from GSoC) I want to add the commands to open the current
document's repository root in the terminal or file explorer too.
>
> Moving the LilyPond data directory here doesn't hurt IMO. When the
"show available fonts" command has been moved to Tools too the "LilyPond"
menu will be exclusively for engraving-related matters and could even
be renamed (if people don't object too much to that ...)

image

---

When the "show available fonts" PR has been merged I want to add a "Text/Markup" submenu below "Musical Transformations". This will initially hold only that single action, but others should be added:

* choose font (for a single in-place command)
* choose document fonts
* Maybe even (feature request) a text editing dialog that makes LilyPond's markup commands available to a GUI text editor.

**Comment by uliska** [comment]
For completeness: After merging the `force-directions` branch there is an additional submenu Tools=>Musical Transformations=>Directions.

When the musical transformations are made available through context menus there should also be actions to *flip* directions.

image

**Comment by dliessi** [comment]
>> Shouldn't also "show available fonts" be moved to the "tools" menu?

> I will do so once #1075 is merged.

Ok, I have just seen (and commented on) #1076.

**Comment by dliessi** [comment]
Two comments on the icons.

1. I think that the Audio-x-generic icon would look better in the menu without the shadow.
2. The music-directions icon is stretched horizontally to a square: it would be better to prepare it as a square from the beginning.

(Both tasks are easily done with Inkscape.)

**Comment by uliska** [comment]
> 1.   I think that the Audio-x-generic icon would look better in the menu without the shadow.

Actually I'm totally unsure how to best proceed with this issue. I did a web search for "Pango Icons Music" and ended up with a list claiming to be Pango icons (which is where Frescobaldi mainly gets its icons from).

> 2.   The music-directions icon is stretched horizontally to a square: it would be better to prepare it as a square from the beginning.

OK, I'*ll have a look.
I'll see how it works out in Inkscape. I have experienced that SVG files modified/saved in Inkscape were not visible in Frescobaldi.

**Comment by dliessi** [comment]
> I have experienced that SVG files modified/saved in Inkscape were not visible in Frescobaldi.

Maybe you only need to ensure they are saved as "pure SVG" instead of "Inkscape SVG" (but I'm not sure).

**Pull request merged**